crypto/bitfield.go

```diff
@@ -9,18 +9,21 @@ type Bitfield struct {
 	len  int
 }
 
-func (bm *Bitfield) extend(nBytes int) {
+func (bf *Bitfield) extend(nBytes int) {
 	// not sure if this is the most efficient way, but it was suggested here:
 	// https://github.com/golang/go/wiki/SliceTricks#extend
-	bm.data = append(bm.data, make([]byte, nBytes)...)
+	bf.data = append(bf.data, make([]byte, nBytes)...)
 }
 
-func (bm Bitfield) set(byteIdx, bitIdx int) {
-	bm.data[byteIdx] |= 1 << bitIdx
+func (bf *Bitfield) set(byteIdx, bitIdx int) {
+	if !bf.isSet(byteIdx, bitIdx) {
+		bf.len++
+	}
+	bf.data[byteIdx] |= 1 << bitIdx
 }
 
-func (bm Bitfield) isSet(byteIdx, bitIdx int) bool {
-	return bm.data[byteIdx]&(1<<bitIdx) != 0
+func (bf Bitfield) isSet(byteIdx, bitIdx int) bool {
+	return bf.data[byteIdx]&(1<<bitIdx) != 0
```

Comment:
The id and index methods below should use shift `>> 3` and `<< 3`. Although not sure if the compiler replaces `*8` and `/8` accordingly...

Reply:
Why?

Reply:
Well, I guess it is only necessary/useful if the compiler doesn't already replace Divide and Multiply operations with shift operations, which are substantially faster. Doing it ourselves makes it fast for sure, independent of compiler optimizations, but I guess it makes the code slightly less readable. (I guess a comment can alleviate the latter.)

Reply:
On second thought, maybe it is better to leave it as is for clarity... For fun, I studied the assembly code generated by these operations. It is actually shorter code with the `/` and `*` operations, but only because of the `%` operation. That is, the simple `* 8` does become a shift `<< 3`. But the `/ 8` doesn't become only a shift operation, when compiled standalone. However, since the function in question has both `/ 8` and `% 8`, it seems to be able to reuse the parts that aren't shift... (but the whole thing is non-trivial to really understand...) The main takeaway is that compilers are smart and we probably should do less manual optimization unless it is demonstrably faster in benchmarks.